airbyte-integrations/bases/base-python/base_python/sdk/streams/core.py

```diff
@@ -0,0 +1,96 @@
+        :return: A generator which yields all the records in this stream. Each record is a dict from properties to their values matching the schema
+         of the stream.
+        """
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        """
+        :return: A dict of the JSON schema representing this stream.
+
+        The default implementation of this method looks for a JSONSchema file with the same name as this stream's "name" property.
+        Override as needed.
+        """
+        # TODO show an example of using pydantic to define the JSON schema, or reading an OpenAPI spec
+        # TODO change to snakecase by default
+        print(self.__class__)
+        print(self.name)
+        return ResourceSchemaLoader(package_name_from_class(self.__class__)).get_schema(self.name)
+
+
+class IncrementalStream(Stream, ABC):
+    @property
+    def cursor_field(self) -> Union[str, List[str]]:
+        """
+        Override to return the name of the default cursor field used by this stream e.g: an API entity might always use created_at as the cursor field.
+        :return: The name of the field used as a cursor. If the cursor is nested, return an array consisting of the path to the cursor.
+        """
+        return []
+
+    @property
+    def source_defined_cursor(self) -> bool:
+        """
+        Override to indicate that the cursor field is custom
+        """
+        return True
+
+    @property
+    @abstractmethod
+    def continuously_save_state(self) -> bool:
+        """
+        Decides whether state is always safe to checkpoint/save at any point.
+
+        If set to true, a state message is output periodically while syncing this stream (which indicates to the process reading from this stream that
+        state should be saved). If set to false, a state message is output only after the stream has been fully read.
+
+        When possible, set this to true as it maximizes the efficiency of reading this stream. For example, if a sync fails halfway through, setting
+        this flag to true will make the stream to read from the latest state. Setting it to false will make it re-extract all the records
+        during the failed sync.
+
+        This flag should only be set to false if records returned from the underlying data source are not returned in ascending order with respect
+        to the cursor field e.g: if the source does not support reading records in ascending order of created_at date. In those cases, state must
+        only be saved once the full stream has been read, and so this flag should be set to false.
+        """
+
+    @abstractmethod
+    def get_updated_state(self, current_state: Mapping[str, Any], latest_record: Mapping[str, Any]):
+        """
+        Inspect the latest record extracted from the data source and the current state object and return an updated state object.
+        It is safe to mutate the input state object and return it.
```

Comment:
Why would you want to allow that? We shouldn't encourage that.

Reply:
especially taking into account that `Mapping` is immutable (for modification there is `MutableMapping`)
also return type is missing.

Reply:
the alternative is to copy the input state every time (or to create a new state object in this method) which seems not great for performance?